src/components/QueryEditor.tsx

```diff
@@ -0,0 +1,217 @@
+
+const tagUrlKeyLookup = _.invert(tagKeyLookup);
+const tagOptions = _(tagKeyLookup)
+  .map((value, key) => ({
+    label: `${value} (${key})`,
+    value,
+  }))
+  .sortBy('label')
+  .value();
+
+export function QueryEditor({ query, onChange, onRunQuery, datasource }: Props) {
+  const [tagName, setTagName] = useState<string | null>(null);
+  const [tagValue, setTagValue] = useState('');
+  const theme = useTheme2();
+
+  const debouncedOnRunQuery = useMemo(() => _.debounce(onRunQuery, 500), [onRunQuery]);
+
+  const onQueryTextChange = (event: ChangeEvent<HTMLInputElement>) => {
+    onChange({ ...query, queryText: event.target.value });
+    debouncedOnRunQuery();
+  };
+
+  const onRealmChange = (event: ChangeEvent<HTMLInputElement>) => {
+    onChange({ ...query, realm: event.target.value });
+    debouncedOnRunQuery();
+  };
+
+  const onMetricChange = (event: ChangeEvent<HTMLInputElement>) => {
+    onChange({ ...query, metric: event.target.value });
+    debouncedOnRunQuery();
+  };
+
+  const onAliasChange = (event: ChangeEvent<HTMLInputElement>) => {
+    onChange({ ...query, alias: event.target.value });
+    debouncedOnRunQuery();
+  };
+
+  const onTagNameChange = (option: SelectableValue<string>) => {
+    if (option.value) {
+      setTagName(option.value);
+    }
+  };
+
+  const onTagValueChange = (event: ChangeEvent<HTMLInputElement>) => {
+    setTagValue(event.target.value);
+  };
+
+  const onAddTag = () => {
+    if (tagName) {
+      onChange({ ...query, tags: { ...query.tags, [tagKeyLookup[tagName] ?? tagName]: tagValue } });
+      setTagName(null);
+      setTagValue('');
+      debouncedOnRunQuery();
+    }
+  };
+
+  const onRemoveTag = (name: string) => {
+    onChange({ ...query, tags: _.omit(query.tags, name) });
+    debouncedOnRunQuery();
+  };
+
+  const onDownsamplerChange = (option: SelectableValue<string>) => {
+    onChange({ ...query, ...(option.value !== DEFAULT_DOWNSAMPLER && { downsampler: option.value }) });
+    debouncedOnRunQuery();
+  };
+
+  const { queryText, realm, metric, tags, downsampler, alias } = query;
+  const areControlsDisabled = Boolean(queryText);
+  const metricDefinitionsUrl = `${datasource.options.dashboardUrl}/${realm}/metricdefinitions`;
+
+  return (
+    <div>
+      <InlineField label="Realm" labelWidth={LABEL_WIDTH} required invalid={!realm} error="Realm is required">
+        <Input onChange={onRealmChange} value={realm} width={30} />
+      </InlineField>
+      <InlineField label="Alias" labelWidth={LABEL_WIDTH}>
+        <Input onChange={onAliasChange} value={alias} width={30} />
+      </InlineField>
+      <InlineField
+        label="Query Text"
+        labelWidth={LABEL_WIDTH}
+        tooltip="If you know the metrics API query
+        string you can paste it here, e.g. m=tasks&u=bob&downsampler=maximum. Otherwise, use the fields
+        below to build your query."
+      >
+        <Input onChange={onQueryTextChange} value={queryText || ''} width={60} />
```

Comment:
queryText sounds like it should be a description. Pepperdata Query? I think this input should go at the end. Maybe even separated by an HR

Reply:
What about "Metrics Query String"? I feel like Pepperdata Query is too generic